// @infura/sdk/dist/src/lib/Auth/availableChains.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainUrls = exports.getChainName = exports.Chains = exports.availableChains = void 0;
var Chains = {
    mainnet: 1,
    ropsten: 3,
    rinkeby: 4,
    goerli: 5,
    polygon: 137,
    mumbai: 80001,
    sepolia: 11155111,
    bsc: 56,
    bsctest: 97,
    avalanche: 43114,
    avalanchetest: 43113,
    fantom: 250,
    cronos: 25,
    cronostestnet: 338,
    arbitrum: 42161,
    palm: 11297108109,
};
exports.Chains = Chains;
var chainsName = {
    1: 'mainnet',
    3: 'ropsten',
    4: 'rinkeby',
    5: 'goerli',
    137: 'polygon',
    80001: 'mumbai',
    11155111: 'sepolia',
    56: 'bsc',
    97: 'bsc test',
    43114: 'avalance',
    43113: 'avalanche test',
    250: 'fantom',
    25: 'cronos',
    338: 'cronos test',
    11297108109: 'palm',
    42161: 'arbitrum',
};
var chainUrls = (_a = {},
    _a[Chains.mainnet] = 'https://mainnet.infura.io',
    _a[Chains.ropsten] = 'https://ropsten.infura.io',
    _a[Chains.rinkeby] = 'https://rinkeby.infura.io',
    _a[Chains.goerli] = 'https://goerli.infura.io',
    _a[Chains.polygon] = 'https://polygon-mainnet.infura.io',
    _a[Chains.mumbai] = 'https://polygon-mumbai.infura.io',
    _a[Chains.sepolia] = 'https://sepolia.infura.io',
    _a[Chains.avalanche] = 'https://avalanche-mainnet.infura.io',
    _a[Chains.avalanchetest] = 'https://avalanche-fuji.infura.io',
    _a[Chains.palm] = 'https://palm-mainnet.infura.io',
    _a[Chains.arbitrum] = 'https://arbitrum-mainnet.infura.io',
    _a);
exports.chainUrls = chainUrls;
var availableChains = [
    Chains.mainnet,
    Chains.goerli,
    Chains.ropsten,
    Chains.rinkeby,
    Chains.polygon,
    Chains.mumbai,
    Chains.sepolia,
    Chains.arbitrum,
    Chains.avalanche,
    Chains.avalanchetest,
    Chains.bsc,
    Chains.bsctest,
    Chains.cronos,
    Chains.cronostestnet,
    Chains.fantom,
    Chains.palm,
];
exports.availableChains = availableChains;
var getChainName = function (chainId) { return chainsName[chainId]; };
exports.getChainName = getChainName;
