// @infura/sdk/dist/src/lib/ContractComponents/hasAccessControl.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ethers_1 = require("ethers");
var Logger_1 = require("../Logger");
var utils_1 = require("../utils");
var constants_1 = require("../constants");
var HasAccessControl = /** @class */ (function () {
    function HasAccessControl() {
    }
    HasAccessControl.prototype.setContract = function (contract) {
        this.contractDeployed = contract;
        this.contractAddress = contract.address;
    };
    /**
     * Add minter function: Grant the 'minter' role to an address
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address to be elevated at 'minter' role
     * @notice Warning: This method will consume gas (30000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.addMinter = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ADDMINTER,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ADDMINTER,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.grantRole(constants_1.DEFAULT_MINTER_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_ADDMINTER,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Renounce minter function: Renounce the 'minter' role
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address that will renounce its 'minter' role
     * @notice Warning: This method will consume gas (40000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.renounceMinter = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEMINTER,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEMINTER,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.renounceRole(constants_1.DEFAULT_MINTER_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEMINTER,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Remove minter function: Remove the 'minter' role to an address
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address that will loose the 'minter' role
     * @notice Warning: This method will consume gas (30000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.removeMinter = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEMINTER,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEMINTER,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.revokeRole(constants_1.DEFAULT_MINTER_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEMINTER,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Is minter function: Check if an address has the 'minter' role or not
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address to check
     * @returns {Promise<boolean>} Promise that will return a boolean
     */
    HasAccessControl.prototype.isMinter = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ISMINTER,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ISMINTER,
                    });
                }
                try {
                    return [2 /*return*/, this.contractDeployed.hasRole(constants_1.DEFAULT_MINTER_ROLE, params.publicAddress)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_ISMINTER,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Add Admin function: Add the 'admin' role to an address. Only callable by
     * addresses with the admin role.
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address that will loose the 'minter' role
     * @notice Warning: This method will consume gas (30000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.addAdmin = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ADDADMIN,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ADDADMIN,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.grantRole(constants_1.DEFAULT_ADMIN_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_ADDADMIN,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Remove Admin function: Remove the 'admin' role to an address. Only callable by
     * addresses with the admin role.
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address that will loose the 'minter' role
     * @notice Warning: This method will consume gas (40000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.removeAdmin = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEADMIN,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEADMIN,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.revokeRole(constants_1.DEFAULT_ADMIN_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_REMOVEADMIN,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Renounce Admin function: Remove the 'admin' role to an address. Only callable by
     * address invoking the request.
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address that will loose the 'minter' role
     * @notice Warning: This method will consume gas (30000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.renounceAdmin = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEADMIN,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEADMIN,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.renounceRole(constants_1.DEFAULT_ADMIN_ROLE, params.publicAddress, options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEADMIN,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Is Admin function: Check whether an address has the 'admin' role
     * @param {object} params object containing all parameters
     * @param {string} params.publicAddress the address to check
     * @returns {Promise<boolean>} Promise that will return a boolean
     */
    HasAccessControl.prototype.isAdmin = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this.contractDeployed && !this.contractAddress) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ISADMIN,
                    });
                }
                if (!params.publicAddress || !ethers_1.ethers.utils.isAddress(params.publicAddress)) {
                    Logger_1.log.throwMissingArgumentError(Logger_1.Logger.message.invalid_public_address, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_ISADMIN,
                    });
                }
                try {
                    return [2 /*return*/, this.contractDeployed.hasRole(constants_1.DEFAULT_ADMIN_ROLE, params.publicAddress)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_ISADMIN,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    /**
     * Renouncing ownership of the smart contract (will leave the contract without an owner).
     * @notice Warning: This method will consume gas (25000 gas estimated)
     * @returns {Promise<ethers.providers.TransactionResponse>} Transaction
     */
    HasAccessControl.prototype.renounceOwnership = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options;
            return __generator(this, function (_a) {
                if (!this.contractAddress && !this.contractDeployed) {
                    Logger_1.log.throwArgumentError(Logger_1.Logger.message.contract_not_deployed, 'contractAddress', this.contractAddress, {
                        location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEOWNERSHIP,
                    });
                }
                try {
                    options = (0, utils_1.addGasPriceToOptions)({}, params.gas);
                    return [2 /*return*/, this.contractDeployed.renounceOwnership(options)];
                }
                catch (error) {
                    return [2 /*return*/, Logger_1.log.throwError(Logger_1.Logger.message.ethers_error, Logger_1.Logger.code.NETWORK, {
                            location: Logger_1.Logger.location.HASACCESSCONTROL_RENOUNCEOWNERSHIP,
                            error: error,
                        })];
                }
                return [2 /*return*/];
            });
        });
    };
    return HasAccessControl;
}());
exports.default = HasAccessControl;
